### PR TITLE
Add rehash for multiset and multimap

### PR DESCRIPTION
**Pull request by PointKernel**
This PR adds `rehash` APIs for multiset and multimap. It also adds `size` to multimap.

### CODE REVIEW

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 1]
```diff
@@ -398,6 +398,79 @@ static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Stora
...
+          class Extent,
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::rehash(
+  cuda::stream_ref stream)
+{
+  this->impl_->rehash(*this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap.inl]:
```suggestion
  impl_->rehash(*this, stream);
```

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 2]
```diff
@@ -398,6 +398,79 @@ static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Stora
...
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::rehash(
+  size_type capacity, cuda::stream_ref stream)
+{
+  auto const extent = make_bucket_extent<static_multimap>(capacity);
+  this->impl_->rehash(extent, *this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap.inl]:
```suggestion
  impl_->rehash(extent, *this, stream);
```

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 3]
```diff
@@ -398,6 +398,79 @@ static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Stora
...
+          class Extent,
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::
+  rehash_async(cuda::stream_ref stream)
+{
+  this->impl_->rehash_async(*this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap.inl]:
```suggestion
  impl_->rehash_async(*this, stream);
```

File: include/cuco/detail/static_multimap/static_multimap.inl [hunk 4]
```diff
@@ -398,6 +398,79 @@ static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Stora
...
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multimap<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::
+  rehash_async(size_type capacity, cuda::stream_ref stream)
+{
+  auto const extent = make_bucket_extent<static_multimap>(capacity);
+  this->impl_->rehash_async(extent, *this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multimap/static_multimap.inl]:
```suggestion
  impl_->rehash_async(extent, *this, stream);
```

File: include/cuco/detail/static_multiset/static_multiset.inl [hunk 5]
```diff
@@ -474,6 +474,60 @@ static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>
...
+          class Extent,
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::rehash(
+  cuda::stream_ref stream)
+{
+  this->impl_->rehash(*this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multiset/static_multiset.inl]:
```suggestion
  impl_->rehash(*this, stream);
```

File: include/cuco/detail/static_multiset/static_multiset.inl [hunk 6]
```diff
@@ -474,6 +474,60 @@ static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>
...
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::rehash(
+  size_type capacity, cuda::stream_ref stream)
+{
+  auto const extent = make_bucket_extent<static_multiset>(capacity);
+  this->impl_->rehash(extent, *this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multiset/static_multiset.inl]:
```suggestion
  impl_->rehash(extent, *this, stream);
```

File: include/cuco/detail/static_multiset/static_multiset.inl [hunk 7]
```diff
@@ -474,6 +474,60 @@ static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>
...
+          class Extent,
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::rehash_async(
+  cuda::stream_ref stream)
+{
+  this->impl_->rehash_async(*this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multiset/static_multiset.inl]:
```suggestion
  impl_->rehash_async(*this, stream);
```

File: include/cuco/detail/static_multiset/static_multiset.inl [hunk 8]
```diff
@@ -474,6 +474,60 @@ static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>
...
+          cuda::thread_scope Scope,
+          class KeyEqual,
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+void static_multiset<Key, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::rehash_async(
+  size_type capacity, cuda::stream_ref stream)
+{
+  auto const extent = make_bucket_extent<static_multiset>(capacity);
+  this->impl_->rehash_async(extent, *this, stream);
```
Comment (sleeepyjack) [inline include/cuco/detail/static_multiset/static_multiset.inl]:
```suggestion
  impl_->rehash_async(extent, *this, stream);
```

File: include/cuco/static_multimap.cuh [hunk 9]
```diff
@@ -602,6 +602,71 @@ class static_multimap {
...
+   *
+   * @note Changes the number of slots to a value that is not less than `capacity`, then
+   * rehashes the container, i.e. puts the elements into appropriate slots considering
+   * that the total number of slots has changed.
+   *
+   * @note This function synchronizes the given stream. For asynchronous execution use
+   * `rehash_async`.
+   *
+   * @note Behavior is undefined if the desired `capacity` is insufficient to store all of the
+   * contained elements.
```
Comment (sleeepyjack) [inline include/cuco/static_multimap.cuh:626]: Note: for the syncronized version we should throw an exception in this case.

**Review by sleeepyjack — APPROVED** [review]

File: include/cuco/static_multimap.cuh (hunk 9 above)
Comment (PointKernel) [reply include/cuco/static_multimap.cuh:626]: How can we verify this without directly calculating the number of filled slots and comparing it to the total capacity?

**Pull request merged**